### PR TITLE
Simplify inline module data handling

### PR DESCRIPTION
**Pull request by JakePT**
## Summary

Addresses issue:

- #11120 

## Relevant technical choices

- Removes the `Module_With_Inline_Data_Trait` trait and updates the `Module_With_Inline_Data::get_inline_data()` interface to return the module's data, rather than adding it to data passed by a filter.
- Updates `Modules::inline_modules_data()` to add inline data from all modules using `Module_With_Inline_Data`. The implementation ensures that any existing data added by the filter is merged into the data from modules, such as the data added by `Ads::register_persistent()`.
- Updates `Ads`, `Analytics_4` and `Sign_In_With_Google` to use the updated interface.
- Updates tests to use the `get_inline_data()` method instead of the filter.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.
- [x] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

### CODE REVIEW

File: includes/Core/Modules/Modules.php [hunk 1]
```diff
@@ -492,7 +506,7 @@ function ( Module $module ) {
 	 *
 	 * @since 1.0.0
 	 *
-	 * @return array Active modules as $slug => $module pairs.
+	 * @return Module[] Active modules as $slug => $module pairs.
```
Comment (JakePT) [inline includes/Core/Modules/Modules.php:509]: The static analysis test was failing without this change.

File: tests/phpunit/integration/Core/Modules/ModulesTest.php [hunk 2]
```diff
@@ -1271,4 +1271,45 @@ public function test_get_feature_metrics__multiple_and_disconnected_shared_modul
...
+
+		$this->assertArrayHasKey( 'fake-module', $inline_modules_data );
+		$this->assertEquals(
+			array(
+				'testInlineData' => true,
+				'existingData'   => true,
+			),
+			$inline_modules_data['fake-module']
+		);
+	}
```
Comment (JakePT) [inline tests/phpunit/integration/Core/Modules/ModulesTest.php:1314]: I'd really appreciate feedback on this. I'm not really sure at this stage what the expectations or preferences are for tests of this kind. Should I have used the actual modules? Or mocked modules with bad or missing inline data? That sort of thing.

Comment (techanvil) [reply tests/phpunit/integration/Core/Modules/ModulesTest.php:1314]: The approach you've taken here, implementing the interface in `FakeModule`, is fine - it does the job well and is consistent with what's already there.

Also the test coverage is OK; you could add some additional cases as you've suggested, but we don't have a strict requirement for 100% coverage, we try to be pragmatic and focus our efforts where the tests will provide most value. I think this will suffice here but please do test more edge cases where it seems sensible to do so.

**Review by techanvil — APPROVED** [review]
LGTM, nice work @JakePT!

✅ 

Note that I've confirmed the failing CI checks are unrelated to this PR.
